### PR TITLE
[Model Support] FLUX.1-dev

### PR DESCRIPTION
**Pull request by raoulritter**

### CODE REVIEW

File: python/src/diffusionkit/mlx/model_io.py [hunk 1]
```diff
@@ -710,6 +718,13 @@ def load_flux(
             hidden_size=config.hidden_size,
             mlp_ratio=config.mlp_ratio,
         )
+    elif model_key == "FLUX.1-dev":
```
Comment (atiorh) [inline python/src/diffusionkit/mlx/model_io.py]: Shall we consolidate this under `if model_key in ["FLUX.1-schnell", "FLUX.1-dev"]` and dedupe code?

File: python/src/diffusionkit/mlx/__init__.py [hunk 2]
```diff
@@ -39,12 +39,14 @@
...
     "FLUX.1-schnell": "argmaxinc/mlx-FLUX.1-schnell",
     "FLUX.1-schnell-4bit-quantized": "argmaxinc/mlx-FLUX.1-schnell-4bit-quantized",
+    "FLUX.1-dev": "raoulritter/flux-dev-mlx",
 }
 
 T5_MAX_LENGTH = {
     "stable-diffusion-3-medium": 512,
     "FLUX.1-schnell": 256,
     "FLUX.1-schnell-4bit-quantized": 256,
+    "FLUX.1-dev": 256,
```
Comment (atiorh) [inline python/src/diffusionkit/mlx/__init__.py]: I believe this should be 512?

File: python/test-gen.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
+import os
```
Comment (atiorh) [inline python/test-gen.py]: Let's remove this file  🙏

File: python/src/diffusionkit/mlx/test-conversion-mlx.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
+import mlx.core as mx
```
Comment (atiorh) [inline python/src/diffusionkit/mlx/test-conversion-mlx.py]: Could you please follow the `UnitTest.TestCase` usage and make this a simple unit test like [this](https://github.com/argmaxinc/DiffusionKit/blob/main/python/src/diffusionkit/tests/mlx/test_diffusion_pipeline.py#L33)? Also, no need to upload to hub from within the test 👍

**Review by atiorh — COMMENTED** [review]
Thanks @raoulritter! Did you test `FLUX.1-schnell` after your changes?

**Comment by raoulritter** [comment]
Hey @atiorh, I've made the changes as requested. Just unsure in regards to what you exactly expect from the test-conversion-mlx? Should it be moved to the tests or should it maybe even be deleted as it's not really needed.


Also checking the merge conflicts it appears these are related to the hosting of the checkpoint. Assuming your team will host these checkpoints yourself so won't resolve these for you. If there is anything else let me know.

**Comment by atiorh** [comment]
@raoulritter Thanks for the quick iteration! We will revise the testing files, download the ckpt and host it alongside schnell. cc: @arda-argmax

**Comment by arda-argmax** [comment]
Hey @raoulritter, I have encountered with problems while trying to run your code. I think you need to make your huggingface checkpoint repo public because I cannot access it. Also, it seems that your branch diverged from our main branch. Can you also rebase it? 🙏 

This is the cli command I'm trying to run:
```
diffusionkit-cli --prompt "a photo of a cat" --output-path out.png --height 512 --width 512 --model-version FLUX.1-dev --shift 1
```

**Comment by raoulritter** [comment]
Hey @arda-argmax,

I think it would perhaps be better based on the conflicts if you guys fixed them according to how you want them instead of me rebasing.  


>I have encountered with problems while trying to run your code. I think you need to make your huggingface checkpoint repo public because I cannot access it`


I previously wrote the following

> Also checking the merge conflicts it appears these are related to the hosting of the checkpoint. Assuming your team will host these checkpoints yourself so won't resolve these for you. 


Let me know if you still want me to take a look but perhaps would make more sense from your side.

**Comment by arda-argmax** [comment]
I see, sorry I missed it. Can you also share how you run the dev model?

I'm getting this error when I'm trying to run your branch with the following command:
```cli
DiffusionKit % diffusionkit-cli --prompt "a photo of a cat" --output-path out.png --height 512 --width 512 --model-version FLUX.1-dev --shift 1
```
Error:
```
scikit-learn version 1.5.1 is not supported. Minimum required version: 0.17. Maximum required version: 1.1.2. Disabling scikit-learn conversion API.
Torch version 2.4.0 has not been tested with coremltools. You may run into unexpected errors. Torch 2.2.0 is the most recent version that has been tested.
WARNING:diffusionkit.mlx.scripts.generate_images:Disabling CFG for FLUX.1-schnell model.
/opt/miniconda3/envs/dkit-test/lib/python3.11/site-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(
INFO:diffusionkit.mlx.scripts.generate_images:Output image resolution will be 512x512
INFO:diffusionkit.mlx:Pre text encoding peak memory: 0.0GB
INFO:diffusionkit.mlx:Pre text encoding active memory: 0.0GB
INFO:diffusionkit.mlx:Post text encoding peak memory: 2.761GB
INFO:diffusionkit.mlx:Post text encoding active memory: 0.478GB
INFO:diffusionkit.mlx:Text encoding time: 0.616s
INFO:diffusionkit.mlx:Pre denoise peak memory: 0.0GB
INFO:diffusionkit.mlx:Pre denoise active memory: 0.004GB
INFO:diffusionkit.mlx:Seed: 1725483544
  0%|                                                                                                                                                                  | 0/50 [00:00<?, ?it/s]INFO:diffusionkit.mlx.mmdit:Cached modulation_params for timesteps=array([1000, 980, 960, ..., 40, 20, 0], dtype=bfloat16)
INFO:diffusionkit.mlx.mmdit:Cached modulation_params will reduce peak memory by 6.5 GB
  0%|                                                                                                                                                                  | 0/50 [00:04<?, ?it/s]
Traceback (most recent call last):
  File "/opt/miniconda3/envs/dkit-test/bin/diffusionkit-cli", line 33, in <module>
    sys.exit(load_entry_point('diffusionkit', 'console_scripts', 'diffusionkit-cli')())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/argmax-mac-studio/arda/test_repos/DiffusionKit/python/src/diffusionkit/mlx/scripts/generate_images.py", line 166, in cli
    image, _ = sd.generate_image(
               ^^^^^^^^^^^^^^^^^^
  File "/Users/argmax-mac-studio/arda/test_repos/DiffusionKit/python/src/diffusionkit/mlx/__init__.py", line 412, in generate_image
    latents, iter_time = self.denoise_latents(
                         ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/argmax-mac-studio/arda/test_repos/DiffusionKit/python/src/diffusionkit/mlx/__init__.py", line 284, in denoise_latents
    latent, iter_time = sample_euler(
                        ^^^^^^^^^^^^^
  File "/Users/argmax-mac-studio/arda/test_repos/DiffusionKit/python/src/diffusionkit/mlx/__init__.py", line 767, in sample_euler
    denoised = model(x, timesteps[i], sigmas[i], **extra_args)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/argmax-mac-studio/arda/test_repos/DiffusionKit/python/src/diffusionkit/mlx/__init__.py", line 703, in __call__
    mmdit_output = self.model.mmdit(**mmdit_input)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/argmax-mac-studio/arda/test_repos/DiffusionKit/python/src/diffusionkit/mlx/mmdit.py", line 222, in __call__
    latent_image_embeddings, token_level_text_embeddings = block(
                                                           ^^^^^^
  File "/Users/argmax-mac-studio/arda/test_repos/DiffusionKit/python/src/diffusionkit/mlx/mmdit.py", line 573, in __call__
    image_intermediates = self.image_transformer_block.pre_sdpa(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/argmax-mac-studio/arda/test_repos/DiffusionKit/python/src/diffusionkit/mlx/mmdit.py", line 444, in pre_sdpa
    modulation_params = self._modulation_params[timestep.item()]

ValueError: [convert] Only length-1 arrays can be converted to Python scalars.
```

File: python/src/diffusionkit/mlx/config.py [hunk 5]
```diff
@@ -90,6 +92,22 @@ def hidden_size(self) -> int:
...
+    depth_unified=38,
+    parallel_mlp_for_unified_blocks=True,
+    hidden_size_override=3072,
+    patchify_via_reshape=True,
+    pos_embed_type=PositionalEncoding.PreSDPARope,
+    rope_axes_dim=(16, 56, 56),
+    pooled_text_embed_dim=768,  # CLIP-L/14 only
+    use_qk_norm=True,
+    float16_dtype=mx.bfloat16,
+    guidance_embed=True,  # Add this line
```
Comment (ZachNagengast) [inline python/src/diffusionkit/mlx/config.py]: This seems to be an unintentional comment?
```suggestion
    guidance_embed=True,
```

File: python/src/diffusionkit/mlx/__init__.py [hunk 6]
```diff
@@ -39,12 +39,14 @@
     "sd3-8b-unreleased": "models/sd3_8b_beta.safetensors",  # unreleased
     "FLUX.1-schnell": "argmaxinc/mlx-FLUX.1-schnell",
     "FLUX.1-schnell-4bit-quantized": "argmaxinc/mlx-FLUX.1-schnell-4bit-quantized",
+    "FLUX.1-dev": "raoulritter/flux-dev-mlx",
```
Comment (ZachNagengast) [inline python/src/diffusionkit/mlx/__init__.py]: Using argmax repo
```suggestion
    "FLUX.1-dev": "argmaxinc/mlx-FLUX.1-dev",
```

File: python/src/diffusionkit/mlx/mmdit.py [hunk 7]
```diff
@@ -209,13 +214,15 @@ def __call__(
         else:
             positional_encodings = None
 
+        timestep_embedding = self.guidance_in(self.t_embedder(timestep))
```
Comment (ZachNagengast) [inline python/src/diffusionkit/mlx/mmdit.py]: Override original timestep only for dev model
```suggestion
        if config.guidance_embed:
            timestep = self.guidance_in(self.t_embedder(timestep))
```

File: python/src/diffusionkit/mlx/mmdit.py [hunk 8]
```diff
@@ -209,13 +214,15 @@ def __call__(
...
+
         # MultiModalTransformer layers
         if self.config.depth_multimodal > 0:
             for bidx, block in enumerate(self.multimodal_transformer_blocks):
                 latent_image_embeddings, token_level_text_embeddings = block(
                     latent_image_embeddings,
                     token_level_text_embeddings,
-                    timestep,
+                    timestep_embedding,
```
Comment (ZachNagengast) [inline python/src/diffusionkit/mlx/mmdit.py]: Override original timestep only for dev model

```suggestion
                    timestep,
```

File: python/src/diffusionkit/mlx/mmdit.py [hunk 9]
```diff
@@ -228,18 +235,17 @@ def __call__(
...
 
             latent_image_embeddings = latent_unified_embeddings[
                 :, token_level_text_embeddings.shape[1] :, ...
             ]
 
-        # Final layer
         latent_image_embeddings = self.final_layer(
             latent_image_embeddings,
-            timestep,
+            timestep_embedding
```
Comment (ZachNagengast) [inline python/src/diffusionkit/mlx/mmdit.py]: Override original timestep only for dev model

```suggestion
            timestep,
```

File: python/src/diffusionkit/mlx/model_io.py [hunk 10]
```diff
@@ -46,6 +46,10 @@
         "FLUX.1-schnell-4bit-quantized": "flux-schnell-4bit-quantized.safetensors",
         "vae": "ae.safetensors",
     },
+    "raoulritter/flux-dev-mlx": {
+        "FLUX.1-dev": "flux1-dev-mlx.safetensors",
+        "vae": "ae.safetensors",
+    },
```
Comment (ZachNagengast) [inline python/src/diffusionkit/mlx/model_io.py]: Use argmax repo
```suggestion
    "argmaxinc/mlx-FLUX.1-dev": {
        "argmaxinc/mlx-FLUX.1-dev": "flux1-dev.safetensors",
        "vae": "ae.safetensors",
    },
```

File: python/src/diffusionkit/mlx/model_io.py [hunk 11]
```diff
@@ -75,6 +79,10 @@
         "vae_encoder": "encoder.",
         "vae_decoder": "decoder.",
     },
+    "raoulritter/flux-dev-mlx": {
+            "vae_encoder": "encoder.",
+            "vae_decoder": "decoder.",
+        },
```
Comment (ZachNagengast) [inline python/src/diffusionkit/mlx/model_io.py]:
```suggestion
    "argmaxinc/mlx-FLUX.1-dev": {
            "vae_encoder": "encoder.",
            "vae_decoder": "decoder.",
        },
```

**Review by ZachNagengast — CHANGES_REQUESTED** [review]
Thanks for contributing, this will be really cool! Tests still need some work to conform to the rest of the repo's style, but added some suggestions to use the argmax HF repo. If we want to follow up with tests in another PR, you may want to just confirm the pipeline is working with the new repo, and take out the current tests for later.

**Comment by ZachNagengast** [comment]
Also updated the merge conflicts 👍 Looks like just a linter issue now

**Comment by ZachNagengast** [comment]
Added a couple changes and got it working locally, just need to fix/remove the tests 👍 
image
```bash
diffusionkit-cli --prompt "detailed cinematic dof render of a \
detailed MacBook Pro on a wooden desk in a dim room with items \
around, messy dirty room. On the screen are the letters 'FLUX on \
DiffusionKit' glowing softly. High detail hard surface render" \
--height 768 \
--width 1360 \
--seed 1001 \
--step 50 \
--output ~/Desktop/flux_on_mac.png \
--model-version argmaxinc/mlx-FLUX.1-dev
```

**Review by arda-argmax — APPROVED** [review]
Everything seems to be working 👍 LGTM!
@raoulritter Thank you for your contribution and @ZachNagengast thank you for the pipeline fixes 🙏

**Comment by mgierschdev** [comment]
Trying to run from this branch and I am getting:
huggingface_hub.utils._errors.RepositoryNotFoundError: 404 Client Error. (Request ID: Root=1-66df5923-68833e15000b3b0d21d61def;e6a2d04b-f799-46e3-ab18-23cf484b0657)

Repository Not Found for url: https://huggingface.co/argmaxinc/mlx-FLUX.1-dev/resolve/main/flux1-dev.safetensors.
Please make sure you specified the correct `repo_id` and `repo_type`.
If you are trying to access a private or gated repo, make sure you are authenticated.

**Comment by ZachNagengast** [comment]
@mgierschdev Thanks for the report, there were some restrictions on the model repo that I've removed now, could you try it again and lmk how it goes? You may still need to accept the terms of the [license](https://huggingface.co/black-forest-labs/FLUX.1-dev) from black forest labs to use it.
image

**Comment by ZachNagengast** [comment]
@raoulritter This is an awesome contribution, thank you!

**Comment by raoulritter** [comment]
Hi @ZachNagengast happy to help thanks for fixing the conflicts and nits. I would love to (help) add lora capabilities as well not sure what your team's plan is.

**Pull request merged**